README.md
# POA-Ethereum Bridge

[![Join the chat at https://gitter.im/poanetwork/poa-bridge](https://badges.gitter.im/poanetwork/poa-bridge.svg)](https://gitter.im/poanetwork/poa-bridge?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![Waffle.io - Columns and their card count](https://badge.waffle.io/poanetwork/poa-bridge.svg?columns=all)](https://waffle.io/poanetwork/poa-bridge)

This is software to be operated by *POA bridge validators* to faciliate proof-of-authority
based briding of POA to tokens on an **another** Ethereum-based blockchain.

The validators work with POA bridge contracts to convert ether on one chain into the same
amount of ERC20 tokens on the other and back.

This software works in conjunction with other projects:

* [POA Bridge UI](https://github.com/poanetwork/bridge-ui)
* [POA Bridge Smart Contracts](https://github.com/poanetwork/poa-bridge-contracts)
* [POA Bridge Monitoring service](https://github.com/poanetwork/bridge-monitor)
* [POA Bridge Deployment scripts](https://github.com/poanetwork/deployment-bridge)

### Functionality

The bridge connects two chains (`home` and `foreign`). When a user deposits ether into the
bridge contract contract on `home` they get the same amount of ERC20 tokens on `foreign`,
and they can convert them back as well.

#### Deposit

![deposit](./res/deposit.png)

#### Withdraw

![withdraw](./res/withdraw.png)

### How to build 

Requires `rust` and `cargo`: [installation instructions.](https://www.rust-lang.org/en-US/install.html)

Requires `solc` to be in `$PATH`: [installation instructions.](https://solidity.readthedocs.io/en/develop/installing-solidity.html)

Assuming you've cloned the bridge (`git clone git@github.com:poanetwork/poa-bridge.git`), run

```
cd poa-bridge
make
```

and install `../target/release/bridge` in your `$PATH`.

### Running

```
bridge --config config.toml --database db.toml
```

- `--config` - location of the configuration file. configuration file must exist
- `--database` - location of the database file.

Bridge forces TLS for RPC connections by default. However, in some limited scenarios (like local testing),
this might be undesirable. In this case, you can use `--allow-insecure-rpc-endpoints` option to allow non-TLS
endpoints to be used. Ensure, however, that this option is not going to be used in production.

#### Exit Status Codes

| Code | Meaning              |
|------|----------------------|
|    0 | Success              |
|    1 | Unknwon error        |
|    2 | I/O error            |
|    3 | Shutdown requested   |
|    4 | Insufficient funds   |
|    5 | Gas too low          |
|    6 | Gas price is too low |
|    7 | Nonce reused         |
|   10 | Cannot connect       |
|   11 | Connection lost      |
|   12 | Bridge crashed       |
|   20 | RPC error            |

### Configuration [file example](./examples/config.toml)

```toml
keystore = "/path/to/keystore"

[home]
account = "0x006e27b6a72e1f34c626762f3c4761547aff1421"
rpc_host = "http://localhost"
rpc_port = 8545
required_confirmations = 0
password = "home_password.txt"
gas_price_oracle_url = "https://gasprice.poa.network"
gas_price_speed = "instant"
default_gas_price = 10_000_000_000 # 10 GWEI

[foreign]
account = "0x006e27b6a72e1f34c626762f3c4761547aff1421"
rpc_host = "http://localhost"
rpc_port = 9545
required_confirmations = 0
password = "foreign_password.txt"

[authorities]
required_signatures = 2

[transactions]
deposit_confirm = { gas = 3000000 }
deposit_relay = { gas = 3000000 }
withdraw_relay = { gas = 3000000 }
```

#### Options

- `keystore` - path to a keystore directory with JSON keys  

#### home/foreign options

- `home/foreign.account` - authority address on the home (**required**)
- `home/foreign.rpc_host` - RPC host (**required**)
- `home/foreign.rpc_port` - RPC port (**defaults to 8545**)
- `home/foreign.required_confirmations` - number of confirmation required to consider transaction final on home (default: **12**)
- `home/foreign.poll_interval` - specify how often home node should be polled for changes (in seconds, default: **1**)
- `home/foreign.request_timeout` - specify request timeout (in seconds, default: **3600**)
- `home/foreign.password` - path to the file containing a password for the validator's account (to decrypt the key from the keystore)
- `home/foreign.gas_price_oracle_url` - the URL used to query the current gas-price for the home and foreign nodes, this service is known as the gas-price Oracle. This config option defaults to `None` if not supplied in the User's config TOML file. If this config value is `None`, no Oracle gas-price querying will occur, resulting in the config value for `home/foreign.default_gas_price` being used for all gas-prices.
- `home/foreign.gas_price_timeout` - the number of seconds to wait for an HTTP response from the gas price oracle before using the default gas price. Defaults to `10 seconds`.
- `home/foreign.gas_price_speed` - retrieve the gas-price corresponding to this speed when querying from an Oracle. Defaults to `fast`. The available values are: "instant", "fast", "standard", and "slow".
- `home/foreign.default_gas_price` - the default gas price (in WEI) used in transactions with the home or foreign nodes. The `default_gas_price` is used when the Oracle cannot be reached. The default value is `15_000_000_000` WEI (ie. 15 GWEI).
- `home/foreign.concurrent_http_requests` - the number of concurrent HTTP requests allowed in-flight (default: **64**)

#### authorities options

- `authorities.account` - all authorities (**required**)
- `authorities.required_signatures` - number of authorities signatures required to consider action final (**required**)

#### transaction options

- `transaction.deposit_confirm.gas` - specify how much gas should be consumed for each call to the Home contract's `submitSignature()` function
- `transaction.deposit_relay.gas` - specify how much gas should be consumed by deposit relay
- `transaction.withdraw_relay.gas` - specify how much gas should be consumed by withdraw relay

### Database file format

```toml
home_contract_address = "0x49edf201c1e139282643d5e7c6fb0c7219ad1db7"
foreign_contract_address = "0x49edf201c1e139282643d5e7c6fb0c7219ad1db8"
checked_deposit_confirm = 120
checked_deposit_relay = 120
checked_withdraw_relay = 121
```

**all fields are required**

- `home_contract_address` - address of the bridge contract on home chain
- `foreign_contract_address` - address of the bridge contract on foreign chain
- `checked_deposit_confirm` - number of the last block for which an authority has confirmed deposit
- `checked_deposit_relay` - number of the last block for which an authority has relayed deposits to the foreign
- `checked_withdraw_relay` - number of the last block for which an authority has relayed withdraws to the home


RELEASE_NOTES.md
# 0.3.0

In this release, dynamic gas pricing has been added
(#85), allowing to fetch gas price from an external
oracle.

This release addresses some major performance issues
with how RPC requests were coordinated. With this release,
RPC requests are finally properly parallelized, leading
to much better overall performance (#94)

It also addresses some potential loss of database updates
under certain conditions (#95)

Along with this, some old configuration options were
deprecated.

# 0.2.1

This release contains a number of bugfixes and a change in handling gas price.
It is no longer set statically but rather dynamically using an external oracle
(see [config example](examples/config.toml))

# 0.2.0

This release, most notably, fixes a condition in which not all logs might be
retrieved from an RPC endpoint, resulting in the bridge not being able to
see all relevant events.

It also improves the performance by introducing concurrent transaction batching.

On the operations side, it'll now print the context of an occurring error
before exiting to help investigating that error.

# 0.1.0

Initial release


bridge/src/config.rs
use std::path::{Path, PathBuf};
use std::fs;
use std::io::Read;
use std::str::FromStr;
use std::time::Duration;
#[cfg(feature = "deploy")]
use rustc_hex::FromHex;
use web3::types::Address;
#[cfg(feature = "deploy")]
use web3::types::Bytes;
use error::{ResultExt, Error, ErrorKind};
use {toml};

const DEFAULT_POLL_INTERVAL: u64 = 1;
const DEFAULT_CONFIRMATIONS: usize = 12;
const DEFAULT_TIMEOUT: u64 = 3600;
const DEFAULT_RPC_PORT: u16 = 8545;
pub(crate) const DEFAULT_CONCURRENCY: usize = 64;
const DEFAULT_GAS_PRICE_SPEED: GasPriceSpeed = GasPriceSpeed::Fast;
const DEFAULT_GAS_PRICE_TIMEOUT_SECS: u64 = 10;
const DEFAULT_GAS_PRICE_WEI: u64 = 15_000_000_000;

/// Application config.
#[derive(Debug, PartialEq, Clone)]
pub struct Config {
	pub home: Node,
	pub foreign: Node,
	pub authorities: Authorities,
	pub txs: Transactions,
	#[cfg(feature = "deploy")]
	pub estimated_gas_cost_of_withdraw: u32,
	pub keystore: PathBuf,
}

impl Config {
	pub fn load<P: AsRef<Path>>(path: P, allow_insecure_rpc_endpoints: bool) -> Result<Config, Error> {
		let mut file = fs::File::open(path).chain_err(|| "Cannot open config")?;
		let mut buffer = String::new();
		file.read_to_string(&mut buffer).expect("TODO");
		Self::load_from_str(&buffer, allow_insecure_rpc_endpoints)
	}

	fn load_from_str(s: &str, allow_insecure_rpc_endpoints: bool) -> Result<Config, Error> {
		let config: load::Config = toml::from_str(s).chain_err(|| "Cannot parse config")?;
		Config::from_load_struct(config, allow_insecure_rpc_endpoints)
	}

	fn from_load_struct(config: load::Config, allow_insecure_rpc_endpoints: bool) -> Result<Config, Error> {
		let result = Config {
			home: Node::from_load_struct(config.home, allow_insecure_rpc_endpoints)?,
			foreign: Node::from_load_struct(config.foreign, allow_insecure_rpc_endpoints)?,
			authorities: Authorities {
				#[cfg(feature = "deploy")]
				accounts: config.authorities.accounts,
				required_signatures: config.authorities.required_signatures,
			},
			txs: config.transactions.map(Transactions::from_load_struct).unwrap_or_default(),
			#[cfg(feature = "deploy")]
			estimated_gas_cost_of_withdraw: config.estimated_gas_cost_of_withdraw,
			keystore: config.keystore,
		};

		Ok(result)
	}
}

#[derive(Debug, PartialEq, Clone)]
pub struct Node {
	pub account: Address,
	#[cfg(feature = "deploy")]
	pub contract: ContractConfig,
	pub request_timeout: Duration,
	pub poll_interval: Duration,
	pub required_confirmations: usize,
	pub rpc_host: String,
	pub rpc_port: u16,
	pub password: PathBuf,
	pub info: NodeInfo,
	pub gas_price_oracle_url: Option<String>,
	pub gas_price_speed: GasPriceSpeed,
	pub gas_price_timeout: Duration,
	pub default_gas_price: u64,
	pub concurrent_http_requests: usize,
}

use std::sync::{Arc, RwLock};
use web3::types::U256;

#[derive(Debug, Clone)]
pub struct NodeInfo {
    pub nonce: Arc<RwLock<U256>>,
}

impl Default for NodeInfo {
	fn default() -> Self {
		NodeInfo {
			nonce: Arc::new(RwLock::new(U256::zero())),
		}
	}
}

impl PartialEq for NodeInfo {
	fn eq(&self, rhs: &Self) -> bool {
		*self.nonce.read().unwrap() == *rhs.nonce.read().unwrap()
	}
}

impl Node {
	fn from_load_struct(node: load::Node, allow_insecure_rpc_endpoints: bool) -> Result<Node, Error> {
		let gas_price_oracle_url = node.gas_price_oracle_url.clone();

		let gas_price_speed = match node.gas_price_speed {
			Some(ref s) => GasPriceSpeed::from_str(s).unwrap(),
			None => DEFAULT_GAS_PRICE_SPEED
		};

		let gas_price_timeout = {
			let n_secs = node.gas_price_timeout.unwrap_or(DEFAULT_GAS_PRICE_TIMEOUT_SECS);
			Duration::from_secs(n_secs)
		};

		let default_gas_price = node.default_gas_price.unwrap_or(DEFAULT_GAS_PRICE_WEI);
		let concurrent_http_requests = node.concurrent_http_requests.unwrap_or(DEFAULT_CONCURRENCY);

		let rpc_host = node.rpc_host.unwrap();

		if !rpc_host.starts_with("https://") {
			if !allow_insecure_rpc_endpoints {
				return Err(ErrorKind::ConfigError(format!("RPC endpoints must use TLS, {} doesn't", rpc_host)).into());
			} else {
				warn!("RPC endpoints must use TLS, {} doesn't", rpc_host);
			}
		}

		let result = Node {
			account: node.account,
			#[cfg(feature = "deploy")]
			contract: ContractConfig {
				bin: {
					let mut read = String::new();
					let mut file = fs::File::open(node.contract.bin)?;
					file.read_to_string(&mut read)?;
					Bytes(read.from_hex()?)
				}
			},
			request_timeout: Duration::from_secs(node.request_timeout.unwrap_or(DEFAULT_TIMEOUT)),
			poll_interval: Duration::from_secs(node.poll_interval.unwrap_or(DEFAULT_POLL_INTERVAL)),
			required_confirmations: node.required_confirmations.unwrap_or(DEFAULT_CONFIRMATIONS),
			rpc_host,
			rpc_port: node.rpc_port.unwrap_or(DEFAULT_RPC_PORT),
			password: node.password,
			info: Default::default(),
			gas_price_oracle_url,
			gas_price_speed,
			gas_price_timeout,
			default_gas_price,
			concurrent_http_requests,
		};

		Ok(result)
	}

	pub fn password(&self) -> Result<String, Error> {
		use std::io::Read;
		use std::fs;
		let mut f = fs::File::open(&self.password)?;
		let mut s = String::new();
		f.read_to_string(&mut s)?;
		Ok(s.split("\n").next().unwrap().to_string())
	}
}

#[derive(Debug, PartialEq, Default, Clone)]
pub struct Transactions {
	#[cfg(feature = "deploy")]
	pub home_deploy: TransactionConfig,
	#[cfg(feature = "deploy")]
	pub foreign_deploy: TransactionConfig,
	pub deposit_confirm: TransactionConfig,
	pub deposit_relay: TransactionConfig,
	pub withdraw_relay: TransactionConfig,
}

impl Transactions {
	fn from_load_struct(cfg: load::Transactions) -> Self {
		Transactions {
			#[cfg(feature = "deploy")]
			home_deploy: cfg.home_deploy.map(TransactionConfig::from_load_struct).unwrap_or_default(),
			#[cfg(feature = "deploy")]
			foreign_deploy: cfg.foreign_deploy.map(TransactionConfig::from_load_struct).unwrap_or_default(),
			deposit_confirm: cfg.deposit_relay.map(TransactionConfig::from_load_struct).unwrap_or_default(),
			deposit_relay: cfg.deposit_relay.map(TransactionConfig::from_load_struct).unwrap_or_default(),
			withdraw_relay: cfg.withdraw_relay.map(TransactionConfig::from_load_struct).unwrap_or_default(),
		}
	}
}

#[derive(Debug, PartialEq, Default, Clone)]
pub struct TransactionConfig {
	pub gas: u64,
}

impl TransactionConfig {
	fn from_load_struct(cfg: load::TransactionConfig) -> Self {
		TransactionConfig {
			gas: cfg.gas.unwrap_or_default(),
		}
	}
}

#[cfg(feature = "deploy")]
#[derive(Debug, PartialEq, Clone)]
pub struct ContractConfig {
	pub bin: Bytes,
}

#[derive(Debug, PartialEq, Clone)]
pub struct Authorities {
	#[cfg(feature = "deploy")]
	pub accounts: Vec<Address>,
	pub required_signatures: u32,
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub enum GasPriceSpeed {
    Instant,
    Fast,
    Standard,
    Slow,
}

impl FromStr for GasPriceSpeed {
	type Err = ();

	fn from_str(s: &str) -> Result<Self, Self::Err> {
		let speed = match s {
			"instant" => GasPriceSpeed::Instant,
			"fast" => GasPriceSpeed::Fast,
			"standard" => GasPriceSpeed::Standard,
			"slow" => GasPriceSpeed::Slow,
			_ => return Err(()),
		};
		Ok(speed)
	}
}

impl GasPriceSpeed {
	pub fn as_str(&self) -> &str {
		match *self {
			GasPriceSpeed::Instant => "instant",
			GasPriceSpeed::Fast => "fast",
			GasPriceSpeed::Standard => "standard",
			GasPriceSpeed::Slow => "slow",
		}
	}
}

/// Some config values may not be defined in `toml` file, but they should be specified at runtime.
/// `load` module separates `Config` representation in file with optional from the one used
/// in application.
mod load {
	use std::path::PathBuf;
	use web3::types::Address;

	#[derive(Deserialize)]
	#[serde(deny_unknown_fields)]
	pub struct Config {
		pub home: Node,
		pub foreign: Node,
		pub authorities: Authorities,
		pub transactions: Option<Transactions>,
		#[cfg(feature = "deploy")]
		pub estimated_gas_cost_of_withdraw: u32,
		pub keystore: PathBuf,
	}

	#[derive(Deserialize)]
	#[serde(deny_unknown_fields)]
	pub struct Node {
		pub account: Address,
		#[cfg(feature = "deploy")]
		pub contract: ContractConfig,
		pub request_timeout: Option<u64>,
		pub poll_interval: Option<u64>,
		pub required_confirmations: Option<usize>,
		pub rpc_host: Option<String>,
		pub rpc_port: Option<u16>,
		pub password: PathBuf,
		pub gas_price_oracle_url: Option<String>,
		pub gas_price_speed: Option<String>,
		pub gas_price_timeout: Option<u64>,
		pub default_gas_price: Option<u64>,
		pub concurrent_http_requests: Option<usize>,
	}

	#[derive(Deserialize)]
	#[serde(deny_unknown_fields)]
	pub struct Transactions {
		#[cfg(feature = "deploy")]
		pub home_deploy: Option<TransactionConfig>,
		#[cfg(feature = "deploy")]
		pub foreign_deploy: Option<TransactionConfig>,
		pub deposit_confirm: Option<TransactionConfig>,
		pub deposit_relay: Option<TransactionConfig>,
		pub withdraw_relay: Option<TransactionConfig>,
	}

	#[derive(Deserialize)]
	#[serde(deny_unknown_fields)]
	pub struct TransactionConfig {
		pub gas: Option<u64>,
	}

	#[derive(Deserialize)]
	#[serde(deny_unknown_fields)]
	pub struct ContractConfig {
		pub bin: PathBuf,
	}

	#[derive(Deserialize)]
	pub struct Authorities {
		#[cfg(feature = "deploy")]
		#[serde(default)]
		pub accounts: Vec<Address>,
		pub required_signatures: u32,
	}
}

#[cfg(test)]
mod tests {
	use std::time::Duration;
	#[cfg(feature = "deploy")]
	use rustc_hex::FromHex;
	use super::{Config, Node, Transactions, Authorities};
	#[cfg(feature = "deploy")]
	use super::ContractConfig;
	#[cfg(feature = "deploy")]
	use super::TransactionConfig;
	use super::{DEFAULT_TIMEOUT, DEFAULT_GAS_PRICE_SPEED, DEFAULT_GAS_PRICE_TIMEOUT_SECS, DEFAULT_GAS_PRICE_WEI};

	#[test]
	fn load_full_setup_from_str() {
		let toml = r#"
keystore = "/keys"

[home]
account = "0x1B68Cb0B50181FC4006Ce572cF346e596E51818b"
poll_interval = 2
required_confirmations = 100
rpc_host = "127.0.0.1"
rpc_port = 8545
password = "password"

[foreign]
account = "0x0000000000000000000000000000000000000001"
rpc_host = "127.0.0.1"
rpc_port = 8545
password = "password"

[authorities]
required_signatures = 2

[transactions]
"#;

		#[allow(unused_mut)]
		let mut expected = Config {
			txs: Transactions::default(),
			home: Node {
				account: "1B68Cb0B50181FC4006Ce572cF346e596E51818b".into(),
				poll_interval: Duration::from_secs(2),
				request_timeout: Duration::from_secs(DEFAULT_TIMEOUT),
				required_confirmations: 100,
				rpc_host: "127.0.0.1".into(),
				rpc_port: 8545,
				password: "password".into(),
				info: Default::default(),
				gas_price_oracle_url: None,
				gas_price_speed: DEFAULT_GAS_PRICE_SPEED,
				gas_price_timeout: Duration::from_secs(DEFAULT_GAS_PRICE_TIMEOUT_SECS),
				default_gas_price: DEFAULT_GAS_PRICE_WEI,
				concurrent_http_requests: DEFAULT_CONCURRENCY,
			},
			foreign: Node {
				account: "0000000000000000000000000000000000000001".into(),
				poll_interval: Duration::from_secs(1),
				request_timeout: Duration::from_secs(DEFAULT_TIMEOUT),
				required_confirmations: 12,
				rpc_host: "127.0.0.1".into(),
				rpc_port: 8545,
				password: "password".into(),
				info: Default::default(),
				gas_price_oracle_url: None,
				gas_price_speed: DEFAULT_GAS_PRICE_SPEED,
				gas_price_timeout: Duration::from_secs(DEFAULT_GAS_PRICE_TIMEOUT_SECS),
				default_gas_price: DEFAULT_GAS_PRICE_WEI,
				concurrent_http_requests: DEFAULT_CONCURRENCY,
			},
			authorities: Authorities {
				#[cfg(feature = "deploy")]
				accounts: vec![
				],
				required_signatures: 2,
			},
			keystore: "/keys/".into(),
		};

		let config = Config::load_from_str(toml, true).unwrap();
		assert_eq!(expected, config);
	}

	#[test]
	fn load_minimal_setup_from_str() {
		let toml = r#"
keystore = "/keys/"

[home]
account = "0x1B68Cb0B50181FC4006Ce572cF346e596E51818b"
rpc_host = ""
password = "password"

[foreign]
account = "0x0000000000000000000000000000000000000001"
rpc_host = ""
password = "password"

[authorities]
required_signatures = 2
"#;
		let expected = Config {
			txs: Transactions::default(),
			home: Node {
				account: "1B68Cb0B50181FC4006Ce572cF346e596E51818b".into(),
				poll_interval: Duration::from_secs(1),
				request_timeout: Duration::from_secs(DEFAULT_TIMEOUT),
				required_confirmations: 12,
				rpc_host: "".into(),
				rpc_port: 8545,
				password: "password".into(),
				info: Default::default(),
				gas_price_oracle_url: None,
				gas_price_speed: DEFAULT_GAS_PRICE_SPEED,
				gas_price_timeout: Duration::from_secs(DEFAULT_GAS_PRICE_TIMEOUT_SECS),
				default_gas_price: DEFAULT_GAS_PRICE_WEI,
				concurrent_http_requests: DEFAULT_CONCURRENCY,
			},
			foreign: Node {
				account: "0000000000000000000000000000000000000001".into(),
				poll_interval: Duration::from_secs(1),
				request_timeout: Duration::from_secs(DEFAULT_TIMEOUT),
				required_confirmations: 12,
				rpc_host: "".into(),
				rpc_port: 8545,
				password: "password".into(),
				info: Default::default(),
				gas_price_oracle_url: None,
				gas_price_speed: DEFAULT_GAS_PRICE_SPEED,
				gas_price_timeout: Duration::from_secs(DEFAULT_GAS_PRICE_TIMEOUT_SECS),
				default_gas_price: DEFAULT_GAS_PRICE_WEI,
				concurrent_http_requests: DEFAULT_CONCURRENCY,
			},
			authorities: Authorities {
				#[cfg(feature = "deploy")]
				accounts: vec![
				],
				required_signatures: 2,
			},
			keystore: "/keys/".into(),
		};

		let config = Config::load_from_str(toml, true).unwrap();
		assert_eq!(expected, config);
	}
}
